File: adapters/spring-boot-kotlin/src/main/kotlin/io/bugfab/adapter/Storage.kt
package io.bugfab.adapter

/**
 * Storage backend abstraction.
 *
 * Two implementations ship with the adapter:
 *
 *  * [FileStorage] — same on-disk layout as the Python reference
 *    (`index.json` + per-report `bug-NNN.json` + `bug-NNN.png`).
 *  * [JpaStorage] — Spring Data JPA backed; H2 for tests, profile-
 *    switchable to Postgres / MySQL.
 *
 * The interface is intentionally synchronous. Spring MVC controllers
 * are blocking by default; consumers who want reactive can wrap with
 * `Mono.fromCallable { storage.saveReport(...) }` without changing
 * the contract.
 *
 * Conformance: every method that takes a report id MUST validate the
 * `bug-[A-Za-z]?\d{3,}` shape before reaching the persistence layer
 * (path-traversal guard). The controller does this once at the top of
 * each handler so backend implementations can assume the id is safe.
 */
interface Storage {
    fun saveReport(metadata: Map<String, Any?>, screenshotBytes: ByteArray): String

    fun getReport(reportId: String): BugReportDetail?

    fun listReports(
        filters: Map<String, String>,
        page: Int,
        pageSize: Int,
    ): Pair<List<BugReportSummary>, Int>

    fun getScreenshotBytes(reportId: String): ByteArray?

    fun updateStatus(
        reportId: String,
        status: String,
        fixCommit: String = "",
        fixDescription: String = "",
        by: String = "",
    ): BugReportDetail?

    fun deleteReport(reportId: String): Boolean

    fun bulkCloseFixed(by: String = ""): Int

    fun bulkArchiveClosed(): Int

    fun setGithubLink(reportId: String, issueNumber: Long, issueUrl: String): BugReportDetail?

    /** Status-count aggregate for the four lifecycle states. */
    fun computeStats(): Map<String, Int>
}

File: adapters/phoenix/lib/bug_fab/storage.ex
defmodule BugFab.Storage do
  @moduledoc """
  Storage backend behaviour for Bug-Fab adapters.

  Implementations:

  * `BugFab.Storage.FileStorage` — JSON-on-disk (zero external deps).
  * `BugFab.Storage.EctoStorage` — Postgres / SQLite via Ecto.

  Method signatures intentionally match the Python reference adapter's
  `Storage` ABC. Return values are plain maps (the wire-protocol shape)
  so router code never sees backend-specific structs.
  """

  @type report_id :: String.t()
  @type metadata :: map()
  @type screenshot :: binary()
  @type filters :: map()
  @type page :: pos_integer()
  @type page_size :: pos_integer()
  @type state :: term()

  @callback handle(opts :: keyword()) :: state()

  @callback save_report(state(), metadata(), screenshot()) ::
              {:ok, report_id()} | {:error, term()}

  @callback get_report(state(), report_id()) ::
              {:ok, map()} | {:error, :not_found}

  @callback list_reports(state(), filters(), page(), page_size()) ::
              {:ok, %{items: [map()], total: non_neg_integer()}}

  @callback list_stats(state()) :: %{String.t() => non_neg_integer()}

  @callback get_screenshot(state(), report_id()) ::
              {:ok, binary()} | {:error, :not_found}

  @callback update_status(state(), report_id(), map(), keyword()) ::
              {:ok, map()} | {:error, :not_found}

  @callback set_github_link(state(), report_id(), integer(), String.t()) ::
              {:ok, map()} | {:error, :not_found}

  @callback delete_report(state(), report_id()) :: :ok | {:error, :not_found}

  @callback bulk_close_fixed(state(), keyword()) :: {:ok, non_neg_integer()}

  @callback bulk_archive_closed(state()) :: {:ok, non_neg_integer()}
end
